### ISSUE TITLE
Saving and restoring DDPG agent

### ISSUE THREAD

**Issue by Sumsamkhan** [openai/baselines#162]
Can someone please tell me how to save and load a model in the DDPG implementation?

**Comment by watts4speed** [comment]
I have the same issue

**Comment by xmanatee** [comment]
Same here :(

**Comment by hfbozkurt** [comment]
Hey, you can use tf.train.saver as it is described in here: https://www.tensorflow.org/programmers_guide/saved_model
Or you can return the agent from the function. Keep in mind that, before you return from function, do not forget to finish the episode. Otherwise, you cannot use the environment directly.

**Comment by freeze888** [comment]
agent.reset() initializes all parameters randomly.
I made new function agent.reset_test() and use it instead of agent.reset()\
----------------------------------- in train function()
   saver.restore(sess,path)

            agent.initialize_test(sess)

            sess.graph.finalize()

            agent.reset_test()
            obs = env.reset()
            #if eval_env is not None:
            #    eval_obs = eval_env.reset()
            done = False
            episode_reward = 0.
            episode_step = 0
            episodes = 0
            t = 0

-------------------------------------------------------------------------------------------------------------------
    def reset_test(self):
        # Reset internal state after an episode is complete.
        if self.action_noise is not None:
            self.action_noise.reset()
        '''
        if self.param_noise is not None:
            self.sess.run(self.perturb_policy_ops, feed_dict={
                self.param_noise_stddev: self.param_noise.current_stddev,
            })
        '''

**Comment by keithmgould** [comment]
@freeze888 Not sure I follow - `agent.reset()` is [called after every episode](https://github.com/openai/baselines/blob/master/baselines/ddpg/training.py#L103). I don't understand how this method (or modifying this method) could be the issue?

Also it looks like this method adjusts action/parameter noise in the DDPG class, which should not effect a restore?

**Comment by chow0214** [comment]
If you select the noise type as parameter noise, then there is a variable in agent called `agent.parameter_noise.current_stddev`, which is not a tensor, so when you save the trained agent and do the restoring, this parameter will not be recovered.

To restore your agent, you need to either add this parameter to the session or manually initialize your parameter noise std as your latest one.

**Comment by Daniel451** [comment]
I am still confused: is editing the baseline code directly really the way to do it? In my opinion, this would be a flaw in the general design. Shouldn't there be at least something like a log method that gets called every n steps, so that one could just add saving there?

**Comment by joellutz** [comment]
@keithmgould I tried your solution, but it somehow didn't really work as you mentioned. Did you (or someone else) manage to correctly save & restore the DDPG agent? After all, training a RL algorithm without being able to load the trained model is not really useful in my opinion.

**Comment by iSaran** [comment]
@joellutz I agree. I believe it should be like the [HER implementation](https://github.com/openai/baselines/tree/master/baselines/her/experiment) in which there is support for train (save the best policy, the latest policy etc), play a policy etc. However HER has its own DDPG implementation. It would be probably a good idea to standardize the contributed algorithms in this repo  in order to have similar structure and support for basic features (like save and restore policies).

**Comment by keithmgould** [comment]
Hey all, 

Have not solved the problem, so as a workaround I used a different (non-Baseline) implementation of DDPG. If thats an option for you, check out Patrick Emami's implementation [here](https://github.com/pemami4911/deep-rl/tree/master/ddpg). He also wrote a nice intro to the algorithm, [here](http://pemami4911.github.io/blog/2016/08/21/ddpg-rl.html).  I've found it **easy to save and restore**, and it **trains just fine**. Not sure whats going on with the Baseline version.

**Comment by joellutz** [comment]
@keithmgould Yeah I've already tried Emami's implementation. The code is much cleaner & shorter, but the techniques differ (e.g. no parameter noise, observation & layer normalization, critic-l2-regularization). The fact that the baseline implementation worked quite quickly for my environment & has by far better action space exploration (possibly due to [parameter noise](https://blog.openai.com/better-exploration-with-parameter-noise/)) are strong arguments for baseline.

I tried extending Emami's implementation with the missing techniques, but I couldn't add the adaptive param noise & observation normalization there, as I'm relatively new to tensorflow, tflearn & RL in general. Does anyone know how to do param noise & observation normalization in tflearn (in Emami's implementation)? I don't really know what's going on with the tensorflow code in the baseline implementation to be honest.

**Comment by jramak** [comment]
@chow0214 I saved the noise param in a pkl file, and the neural net weights with tensorflow's Saver. However, it still does not restore exactly where it left off.

**Comment by watts4speed** [comment]
Not having the ability is a really big issue, I agree.  Is there anyway we can get an indication from the OpenAI people about whether this issue is something they intend to fix.  It's also a big deal to switch away from baselines/DDPG given the other stuff they have implemented and I don't want to do so if they are going to fix the problem.

**Comment by joellutz** [comment]
I did some testruns with my environment, trying to save & restore the model. I captured the last state right before training & saving of the model, as well as the selected action right after that (in the new epoch-cycle). So I captured the action and the state this action was based on. Then I terminated the training, and started it again, but this time I restored the model from the previously saved file. At the env.reset() (right at the beginning of the episode) I injected the previously captured state. Then I captured the selected action right at the beginning, which should be exactly the same as the previously captured action, because the state each of them are based on is exactly the same.

The two actions were exactly the same when having no parameter or action noise at all, so I think the **saving & restoring might have worked**. With parameter noise, the two actions weren't exactly the same, but this is maybe due to the randomness of the parameter noise which leads to a slightly different action. (Even though everything is seeded, the random generators are in a different state right at the beginning vs. right after the first epoch-cycle.)

This is how I save & restore the model (all in baselines/ddpg/training.py, slightly modified compared to @keithmgould's [solution](https://github.com/openai/baselines/issues/162#issuecomment-397356482)
```
# ...
with U.single_threaded_session() as sess:
    # Prepare everything.
    
    if restore == True:
        logger.info("Restoring from saved model")
        saver = tf.train.import_meta_graph(savingModelPath + "ddpg_test_model.meta")
        saver.restore(sess, tf.train.latest_checkpoint(savingModelPath))
    else:
        logger.info("Starting from scratch!")
        sess.run(tf.global_variables_initializer()) # this should happen here and not in the agent right?


    agent.initialize(sess)
    sess.graph.finalize()

    # ...

            # in the epoch_cycles loop (after the training of the model)
            # Saving the trained model
            if(saver is not None):
                logger.info("saving the trained model")
                start_time_save = time.time()
                saver.save(sess, savingModelPath + "ddpg_test_model")
                logger.info('runtime saving: {}s'.format(time.time() - start_time_save))

            logger.info('runtime epoch-cycle {0}: {1}s'.format(cycle, time.time() - start_time_cycle))

        mpi_size = MPI.COMM_WORLD.Get_size()
        # ...
```

The logs of my testrun without any param or action noise:

> \# (last rollout step)
> selected (unscaled) action: [-0.012  0.005  0.007 -0.002]
> Training the Agent
> saving the trained model
> runtime saving: 2.47790503502s
> runtime epoch-cycle 0: 722.832417965s
> selected (unscaled) action: **[ 0.001 -0.001  0.009 -0.035]**
> \# (run aborted)
> 
> \# (new run with restore=True and injected state)
> Restoring from saved model
> INFO:tensorflow:Restoring parameters from ~/Documents/saved_models_OpenAI_gym/ddpg_test_model
> selected (unscaled) action: **[ 0.001 -0.001  0.009 -0.035]**

**Comment by jramak** [comment]
@joellutz seems like you included writing the meta graph in the save compared to @keithmgould 's solution. Did it make a difference when restoring from the meta graph? It says here it's not necessary: https://stackoverflow.com/questions/36195454/what-is-the-tensorflow-checkpoint-meta-file

**Comment by joellutz** [comment]
@jramak I don't know if writing the meta graph is really necessary or not, I followed [this tutorial](http://cv-tricks.com/tensorflow-tutorial/save-restore-tensorflow-models-quick-complete-tutorial/) where they include it. I haven't tried if it still works without, as the process of capturing the state etc. is quite tedious for my environment.

**Comment by LaTinta** [comment]
@joellutz Thanks~ Your method is very effective. 
If someone wants to save and load DDPG model, there are some steps:
1. use saver.save() to save tf session;
2. move sess.run(tf.global_variables_initializer()) from initialize in DDPG, to where you want to initializatize model parameters;
3. use saver.restore() to load tf session, and use initialize() to load DDPG model.
If you need to use the model in a service, I suggest you use tf.InteractiveSession() instead of U.single_threaded_session().

**Comment by brendenpetersen** [comment]
@LaTinta This method will not save the information from the `RunningMeanStd` object, correct? So, if `normalize_observations=True` (the default value), then the agent used at evaluation will not be the same as the one in training.

For my own use cases, I've set `normalize_observations=False` to avoid this issue (and because my observation space was already normalized, so it ended up hurting performance anyway). But with PPO1, for example, the `RunningMeanStd` object is always created (there is no setting to turn it off), so I don't think offline evaluation possible without changing the code.

**Comment by r7vme** [comment]
Seems `stable-baselines` DDPG implementation provides ability to save/load.

https://stable-baselines.readthedocs.io/en/master/modules/ddpg.html#example

**Comment by jrjbertram** [comment]
FYI, as of commit 858afa8d7e these approaches no longer work.  ddpg/training.py was removed.  Looking at adapting them for the refactored codebase.

**Comment by jrjbertram** [comment]
I have it working (I think) in the latest codebase...  except that my model performs poorly after loading from a checkpoint.  Used similar approach as described above.  I can't tell if that's due to an error in my code or just a bad training result.  Does anyone have any ideas on how to verify that the model was loaded correctly?

Code changes in this commit (I have a copy of openai baselines embedded in my repo for now):
https://github.com/jrjbertram/jsbsim_rl/commit/6825c0c277e94d24e3ecb1450eef82dda8b5793d

For my testing I'm using the approach of loading a 0-length training session followed by a --play.  

https://github.com/jrjbertram/jsbsim_rl/blob/master/replay.sh

I verified that the code to reload the model is being executed.


Where I'm confused / suspicious is that during training my rollout return and return_history curves look pretty good...  nice logarithmic shape.   When I replay though my actions seem fairly random and I don't seem to be acting in a way that would collect any reward.

**Comment by Sohojoe** [comment]
@jrjbertram are your normalizing your environment? if so, there is a known issue:

> NOTE: At the moment Mujoco training uses VecNormalize wrapper for the environment which is not being saved correctly; so loading the models trained on Mujoco will not work well if the environment is recreated. If necessary, you can work around that by replacing RunningMeanStd by TfRunningMeanStd in baselines/common/vec_env/vec_normalize.py. This way, mean and std of environment normalizing wrapper will be saved in tensorflow variables and included in the model file; however, training is slower that way - hence not including it by default

source = https://github.com/openai/baselines#saving-loading-and-visualizing-models

**Comment by joyce-fang** [comment]
@r7vme The stable-baselines library doesn't seem to solve the RunningMeanStd issue. It changes the normalize_observation default value to False so the RunningMeanStd is not used. When I enable normalize_observation the model does not restore correctly.

**Comment by joyce-fang** [comment]
@r7vme The stable-baselines library doesn't seem to solve the RunningMeanStd issue. It changes the normalize_observation default value to False so the RunningMeanStd is not used. When I enable normalize_observation the model does not restore correctly.

**Comment by DanielTakeshi** [comment]
Has there been an update on how to properly save DDPG models?
